src/main/scala/machines/Internal.scala

```diff
@@ -2,5 +2,40 @@ package machines
 
 import regex._
 import dfa._
+import scala.Conversion
 
-// TODO: Add your code below
+given Conversion[Char, RegularLanguage] = Character(_)
+
+given Conversion[String, RegularLanguage] = str =>
+    str.tail.foldLeft(str.head: RegularLanguage)(Concat(_, _))
+
+extension (lang: RegularLanguage)
+    def ||(other: RegularLanguage) = Union(lang, other)
+    def ~(other: RegularLanguage) = Concat(lang, other)
+    def <*> = Star(lang)
+    def <+> = Concat(lang, Star(lang))
+    def apply(n: Int): RegularLanguage = 
+        if n <= 0 then Epsilon  // only reached if n starts < 1
+        else if n == 1 then lang
+        else Concat(lang, lang{n-1})
+    // BONUS: range repetition -- used like {2,5}, but I have to use (2,5)
+    // since there are multiple arguments
+    def apply(lower: Int, upper: Int): RegularLanguage =
+        if lower > upper then Epsilon  // only reached of lower starts > upper
+        else if lower == upper then lang{lower}
+        else Union(lang{lower}, lang(lower+1, upper))
+
+    def toDFA(using alphabet: Set[Char]) = regexToDFA(lang, alphabet)
+
+given Conversion[RegularLanguage, DFA] = lang =>
+    lang.toDFA(using chars(lang))
+
+def chars(lang: RegularLanguage): Set[Char] =
+    lang match
+        case Empty => Set()
+        case Epsilon => Set()
+        case Character(c) => Set(c)
+        case Union(lang1, lang2) => chars(lang1) ++ chars(lang2)
+        case Concat(lang1, lang2) => chars(lang1) ++ chars(lang2)
```

Comment:
I opted for the `union` operator because of the way the assignment was written, but I like your use of `++` here because it's a common way collections are added together in programming languages. I looked at the Collections article in the Scala docs again on these operators’ differences. It seems like `++` is an operator on the superclass of `Set`, so your code might be more transferable should the need for different ways of organizing characters arise.

I referred to https://docs.scala-lang.org/overviews/collections-2.13/sets.html for the difference between `++` and `union`.